Sources/Plasma/NucleusLib/pnUtils/plCmdParser.cpp

```diff
@@ -0,0 +1,618 @@
+
+    for (loop = 0; loop < defCount; ++loop) {
+        plCmdArgDef def = defs[loop];
+
+        // Check whether this argument is flagged
+        bool flagged = hsCheckFlagBits(def.flags,
+                                       kCmdArgFlagged,
+                                       kCmdArgMask);
+
+        // Disallow names on unflagged arguments
+        ASSERT(flagged || !def.name.IsEmpty());
+
+        // Store the argument data
+        plCmdArgData& arg = fArgArray[loop];
+        arg.def           = def;
+        arg.buffer        = "";
+        arg.nameChars     = def.name.GetSize();
+        arg.isSpecified   = false;
+
+        SetDefaultValue(arg);
+        maxId = std::max(maxId, def.id);
+
+        // Track the number of unflagged arguments
+        if (!flagged) {
+            ++unflaggedCount;
+        }
+    }
+
+
+    // Build the id lookup table
+    size_t idTableSize = std::min(maxId + 1, defCount * 2);
+    fLookupArray.resize(idTableSize);
+
+    for (loop = 0; loop < defCount; ++loop) {
+        if (defs[loop].id < idTableSize) {
+            fLookupArray[defs[loop].id] = loop;
+        }
+    }
+
+
+    // Build the unflagged array
+    size_t unflaggedIndex = 0;
+    fUnflaggedArray.resize(unflaggedCount);
+
+    for (loop = 0; loop < defCount; ++loop) {
+        bool req = hsCheckFlagBits(defs[loop].flags,
+                                   kCmdArgRequired,
+                                   kCmdArgMask);
+
+        if (req) {
+            fUnflaggedArray[unflaggedIndex++] = loop;
+        }
+    }
+
+    fRequiredCount = unflaggedIndex;
+
+    for (loop = 0; loop < defCount; ++loop) {
+        bool flagged = hsCheckFlagBits(defs[loop].flags,
+                                       kCmdArgFlagged,
+                                       kCmdArgMask);
+
+        bool req     = hsCheckFlagBits(defs[loop].flags,
+                                       kCmdArgRequired,
+                                       kCmdArgMask);
+
+        if (!flagged && !req) {
+            fUnflaggedArray[unflaggedIndex++] = loop;
+        }
+    }
+}
+
+
+void plCmdParserImpl::SetDefaultValue(plCmdArgData& arg)
+{
+    uint32_t argType = arg.def.flags & kCmdTypeMask;
+
+    switch (argType) {
+    case kCmdTypeBool:
+        arg.val.boolVal = !hsCheckFlagBits(arg.def.flags,
+                                           kCmdBoolSet,
+                                           kCmdBoolMask);
+        break;
+
+    case kCmdTypeInt:
+        arg.val.intVal = 0;
+        break;
+
+    case kCmdTypeUint:
+        arg.val.uintVal = 0;
+        break;
+
+    case kCmdTypeFloat:
+        arg.val.floatVal = 0.0f;
+        break;
+
+    case kCmdTypeString:
+        arg.val.strVal = "";
+        break;
+
+    DEFAULT_FATAL(argType);
+    }
+}
+
+
+bool plCmdParserImpl::Tokenize(plCmdTokenState* state, std::vector<plString>& strs)
+{
+    bool result = true;
+
+    for (auto it = strs.begin(); result && it != strs.end(); ++it) {
```

Comment:
There's an extra condition on the looping for both of these instances:
`result && it != strs.end()`